src/solver/activ/fwd_0.cpp

```diff
@@ -0,0 +1,217 @@
+    }
+
+    {
+        auto kernel_info         = KernelInfo{};
+        kernel_info.comp_options = build_params.GenerateFor(kbp::OpenCL{});
+
+        kernel_info.l_wk.push_back(256);
+        kernel_info.l_wk.push_back(1);
+        kernel_info.l_wk.push_back(1);
+
+        const auto MAP_RD = read_len / read_unit;
+
+        kernel_info.g_wk.push_back(MAP_RD);
+        kernel_info.g_wk.push_back(packed ? 1 : height);
+        kernel_info.g_wk.push_back(1);
+
+        kernel_info.kernel_file = "MIOpenNeuron.cl";
+        kernel_info.kernel_name = (packed) ? "MIOpenActiveFwdLite" : "MIOpenActiveFwd2DLite";
+
+        result.construction_params.push_back(kernel_info);
+    }
+
+    result.invoker_factory = [=](const std::vector<Kernel>& kernels) {
+        return [=](const Handle& handle, const AnyInvokeParams& raw_params) {
+            decltype(auto) kernel = handle.Run(kernels.front());
+            decltype(auto) params = raw_params.CastTo<miopen::activ::InvokeParams>();
+
+            visit_float(params.x_desc.GetType(), [&](auto as_float) {
+                const auto alpha = as_float(params.alpha);
+                const auto beta  = as_float(params.beta);
+                const auto gamma = as_float(params.gamma);
+
+                if(packed)
+                {
+                    kernel(params.x,
+                           params.y,
+                           gamma,
+                           beta,
+                           alpha,
+                           static_cast<long long>(params.x_offset),
+                           static_cast<long long>(params.y_offset));
+                }
+                else
+                {
+                    const auto x_lens_ = params.x_desc.GetLengths();
+                    const auto y_lens_ = params.y_desc.GetLengths();
+
+                    const auto x_strides = params.x_desc.GetStrides();
+                    const auto y_strides = params.y_desc.GetStrides();
+
+                    const auto x_stride2D = static_cast<unsigned int>(
+                        (x_lens_.size() == 2) ? x_strides[0] : (x_lens_.size() == 3)
+                                                                   ? x_strides[1]
+                                                                   : (x_lens_.size() == 4)
+                                                                         ? x_strides[2]
+                                                                         : x_strides[3]);
+                    const auto y_stride2D = static_cast<unsigned int>(
+                        (y_lens_.size() == 2) ? y_strides[0] : (y_lens_.size() == 3)
+                                                                   ? y_strides[1]
+                                                                   : (y_lens_.size() == 4)
+                                                                         ? y_strides[2]
+                                                                         : y_strides[3]);
```

Comment:
It seems like many computations are invariant here and can be moved out from the Invoker and even out from the Factory. Shall we open a ticket for this optimization?

Reply:
Oh, but they are not invariant. You are used to convolution, I have been too ;) Tensor descriptors are not part of network config in activation.

Reply:
Wow.

Reply:
Took me a while to even consider that possibility.